.cloud-build/execute_changed_notebooks_helper.py

```diff
@@ -336,12 +343,79 @@ def get_changed_notebooks(
 
     return notebooks
 
+def _save_results(results: List[NotebookExecutionResult],
+                  artifacts_bucket: str,
+                  results_file: str):
+
+    artifacts_bucket = artifacts_bucket.replace("gs://", "").split('/')[0]
+
+    # read in existing prior results data
+    print("Reading existing accumulative results ...")
+    rows = []
+    try:
+        client = storage.Client()
+        bucket = client.get_bucket(artifacts_bucket)
+        blob = bucket.blob(str(results_file))
+        blob = blob.download_as_string()
+        contents = blob.decode('utf=8')
+        contents = io.StringIO(contents)
+
+        header = True
+        for row in csv.reader(contents):
+            if header:
+                header = False
+            else:
+                row[1] = float(row[1])
+                row[2] = int(row[2])
+                row[3] = int(row[3])
+                rows.append(row)
+
+    except Exception as e:
+        print('Exception', e)
+    print(rows)
+
+
+    print("Updating accumulative results ...")
+    for result in results:
+        found = False
+        for row in rows:
+            if row[0] == result.path:
+                found = True
+                row[1] = result.duration.total_seconds()
+                if result.is_pass:
+                    row[2] += 1
+                else:
+                    row[3] += 1
+                print(f"updating {result.path}: {row}")
+                break
+
+        if not found:
+            if result.is_pass:
+                passed = 1
+                failed = 0
+            else:
+                failed = 1
+                passed = 0
+            rows.append([result.path, result.duration.total_seconds(), passed, failed])
+            print(f"adding {result.path}")
+
+    print("Saving accumulative results ...")
+    content = "notebook,duration,passed,failed\n"
+    for row in rows:
+        content += str(row).replace('[', '').replace(']', '').replace('"', "").replace("'", '') + '\n'
+
+    client = storage.Client()
+    bucket = client.get_bucket(artifacts_bucket)
+    bucket.blob(str(results_file)).upload_from_string(content, 'text/csv')
```

Comment:
This still has concurrency issues. You can't write to a specific file that multiple running jobs without a global semaphore or locking mechanism.

You should write each result to `{artifacts_bucket}`/build_results/{build_id}.

Do the accumulation over `{artifacts_bucket}`/build_results/ when you need the accumulated results for something.

Reply:
using BUILD_ID by default